Lazy loading delegate profile

andreitr opened the pull request:
Loom [demo](https://www.loom.com/share/8174e571ab964b35a10fff2d57fd1ee1) 👀

What changed 
- implemented `unstable_cache` for expensive queries 
- optimized delegate lookup
- cached delegate queries on the client + server
- optimized mobile / desktop profile dropdown (queries > client)
- optimized voting stats hook to cache more aggressively 
- moved the Delegate Header (queries > client)
- revalidating `delegate` and `delegateStatement` query tags on profile update

jefag reviewed (changes requested):
@andreitr nice job, but there is enough changing here.  Need to break down whats been tested and how, and then also call out the higher risk areas of the site for getting help from others to test.

jefag commented on src/app/api/common/delegates/getDelegateForSCW.ts:
```diff
@@ -15,4 +15,12 @@ async function getDelegateForSCW(address: string) {
+  },
+  [],
+  {
+    revalidate: 3600, // 1 hour cache
```
So - if a user changes their delegate statement, does that bust the cache?

mcgingras replied:
no, but we could manually bust the cache if we add a tag to the options here then call `revalidateTag`
I think we should do this -- otherwise the site will feel broken if you update your statement but it doesn't update for an hour

andreitr replied:
Michael is correct. We will need to manually invalidate the tag whenever the statement is updated. In this case, we are using the delegate for SCW redirection. This can be cached indefinitely since the SCW and EOA pairing will not change.

jefag commented on src/app/delegates/actions.ts:
```diff
@@ -24,17 +24,36 @@ import { createDelegateStatement } from "@/app/api/common/delegateStatement/crea
+  },
+  [],
+  {
+    revalidate: 600, // 10 minute cache
```
1-minute feels better than 10 mins, no?

mcgingras replied:
I think we should add tags to every `unstable_cache` and we should bust the cache on any writes

andreitr replied:
Good call, @mcgingras. Added revalidation to ensure the latest delegate statement is always displayed.

mcgingras commented on src/app/api/common/delegates/getDelegateForSCW.ts:
```diff
@@ -15,4 +15,12 @@ async function getDelegateForSCW(address: string) {
     .catch((error) => console.error(error));
 }
 
-export const fetchDelegateForSCW = cache(getDelegateForSCW);
```
maybe the `cache` is now redundant, but my understanding is that `cache` (from react) is useful for deduping server function requests. For example, if you have a server component that requests `getDelegate` and then renders another server component that requests `getDelegate` (so like a waterfall that requests the same data) then using `cache` from react will dedup the request, similar to how `useQuery` lets you cache client side data fetches in a single user session.

So, my question is... do we still want this? My thought would be that we keep cache, but then use `unstable_cache` on the component side where we are making the request. This way, we can also scope persistent `unstable_cache` to a component, and maybe invalidate that cache in context of what that component is doing. Maybe this doesn't make sense? I'm not sure. Just calling it out in case we want to talk about it more

mcgingras commented on src/app/api/common/delegates/getDelegates.ts:
```diff
@@ -609,5 +609,5 @@ async function getVoterStats(
 }
 
 export const fetchDelegates = cache(getDelegates);
-export const fetchDelegate = cache(getDelegate);
-export const fetchVoterStats = cache(getVoterStats);
```
it's fine to leave `cache` here -- I think it's still valuable for deduping in case there's an area that we don't use data-cache for. It's a react only thing

mcgingras commented on src/app/lib/ENSUtils.ts:
```diff
@@ -1,26 +1,30 @@
-  if (isAddress(nameOrAddress)) {
-    return nameOrAddress;
-  }
+export const ensNameToAddress = unstable_cache(
```
should we move all ens lookups to this call? Typically I use wagmi's ens hooks, but I like that these are cached more persistently. The wagmi hooks cache in a single session (like useQuery) but then we end up hitting the rpc a bunch of times when we might not need to. I like this

andreitr replied:
@mcgingras all reverse lookups should now go through this call.

mcgingras commented on src/components/DelegateStatement/DelegateStatementForm.tsx:
```diff
@@ -51,21 +44,6 @@ export default function DelegateStatementForm({
-  }, [address]);
-
   async function onSubmit(values: DelegateStatementFormValues) {
     if (!agreeCodeConduct) {
```
this is where we maybe want to revalidate cache

mcgingras commented on src/components/Delegates/Delegations/DelegationsContainer.tsx:
```diff
@@ -126,8 +107,19 @@ function DelegationsContainer({
-                  </InfiniteScroll>
+                  </TableBody>
                 </Table>
+                {meta.has_next && (
+                  <div className="text-center my-4">
+                    <button
+                      onClick={loadMore}
+                      disabled={isLoading}
+                      className="px-4 py-2 text-primary text-sm bg-wash hover:bg-wash/80 rounded-lg"
+                    >
+                      {isLoading ? "Loading..." : "Load More"}
+                    </button>
+                  </div>
+                )}
```
no more infinite scroll?

andreitr replied:
@mcgingras This component was buggy. If you check the production OP/ENS, you’ll notice that it automatically requests page data until the entire list loads. I spent at least an hour debugging it, but ultimately decided to use a simple “Load More” button instead.

mcgingras replied:
works for me, I wasn't a huge fan of the infinite loading anyways. Thanks for the context!

mcgingras commented on src/components/Header/DesktopProfileDropDown.tsx:
```diff
@@ -21,66 +20,61 @@ import {
-  const { isLoading, delegate, balance } = useConnectedDelegate();
-  const hasStatement = !!delegate?.statement;
-  const canCreateDelegateStatement = ui.toggle("delegates/edit")?.enabled;
+  // Don't hydrate the component until the user clicks on the profile dropdown
```
what's the thought behind this comment?

andreitr replied:
Don’t load the delegate profile, voting power, or other data until the user clicks the profile button in the header. The previous version unnecessarily loaded the “connected” delegate statement and all related data, even when the component wasn’t displayed.

mcgingras replied:
Makes sense. I'm only speculating because I haven't looked at the page directly, but would this block the rest of the page? I'm thinking if it's not blocking, it could be nice to have this data "warm" so when the user clicks it's already available rather than having to show a loader when the data is fetched after user intent.

But then again, if it's rare that a user clicks into the profile we are fetching and sending so much data over the network that is never used, so it's wasteful and your change here is a nice save.

Just thinking out loud here, I think your change is totally okay.

andreitr commented:
**QA**

**ENS - locally against the prod db + mainnet**
- gassless delegation [txn](https://etherscan.io/tx/0xb288abcba6cad1f83e99d00df39f4cf2f8c3147abde009dee42896286d538ee2#eventlog)
- gassless undelegation [txn](https://etherscan.io/tx/0x615e8561b71d087b39c8ad50631e1f9882ba9ec3f06331c8c3a91ef83973c184)
- created delegate statement and validated cache reload
- delegated back to self [txn](https://etherscan.io/tx/0x614b85312e78fda654e5c2eb355704d662c74585c585613d3668159830a3ddb0)
- tested large delegate loading speed `delegates/0x839395e20bbb182fa440d08f850e6c7a8f6f0780`

**Uniswap - locally against the prod db + mainnet**
- gassless delegation [txn](https://etherscan.io/tx/0x0552d55269a070703ea70521f779b350bb25caab7531682d4e6e7c854f08c007#eventlog)
- gasless undelegation [txn](https://etherscan.io/tx/0x85ffec565dbe6dc0f9cdf74a429be22480ec1376adac2c2f33b0aac7b9bce490)
- updated delegate statement and validated cache reload 
- delegated back to self [txn](https://etherscan.io/tx/0x54f3f96c81a22094ea3ba953738f3d22e8c77668909feef967631cba0711cb6d)
- tested large delegate loading speed `delegates/0xe7925d190aea9279400cd9a005e33ceb9389cc2b`

**Derive - locally against the prod db + mainnet**
- tested large delegate `delegates/0xf68d2bfcecd7895bba05a7451dd09a1749026454`
- tested large delegate `delegates/0x9ba8c70a8fd922e97a4e78c46583742c7d41796c`
- tested large delegate `delegates/0xfb0cc75cdb9ecd0b13fc2be7771bbbc98b73f0a2`
- NO PROD TOKENS to test delegation

**Optimism - locally against the prod db + mainnet**
- tested large delegate `delegates/0x2b888954421b424c5d3d9ce9bb67c9bd47537d12`
- tested large delegate `delegates/0x7b0befc5b043148cd7bd5cfeeef7bc63d28edec0`
- tested full delegation [txn](https://optimistic.etherscan.io/tx/0x5cf5488810dee27408b80dbe0bfedb1ed649100a7ad95cd85c44e06f2dbd7e2e)
- tested undelegation [txn](https://optimistic.etherscan.io/tx/0x4d61efe2916e551ea8f48732ef46a56e5ceabafba6f5e50dc8be90d385c1b01b)

**Scroll - locally against the prod db + mainnet**
- tested delegate statement update `delegates/0x648BFC4dB7e43e799a84d0f607aF0b4298F932DB`
- tested large delegate `delegates/0x85c2c57b349fff684b1122a755145a02dde38f79`
- tested large delegate `delegates/0xf92a3f83209c7747b8b8ad82e3b3672c53b95f12`

andreitr commented:
**BOOST Testnet QA**

**[Deployment](https://agora-next-boost-git-andrei-eng-508-delegate-o-96589a-voteagora.vercel.app/)**

- create / update [delegate statement](https://agora-next-boost-git-andrei-eng-508-delegate-o-96589a-voteagora.vercel.app/delegates/0x1d671d1b191323a38490972d58354971e5c1cd2a)
- delegated [txn](https://optimistic.etherscan.io/tx/0x28bb1f3fd90a765eaf26d074e4f20a89978e530758ac36c0f101d9a6c86de0e8#eventlog)
- created proposal [txn](https://optimistic.etherscan.io/tx/0x60bb92ccc7cbe2f00f4248aea193b3beaea5e5b8fb4a56522520c4eba864c3eb)
- voted [txn](https://optimistic.etherscan.io/tx/0x800d0342c06f8b21aa82130f702d45b2bbac3c2ad355b4ced6898b5fd05b0584)

I was unable to verify the undelegate functionality. The button is displayed incorrectly, even though I delegated to someone else. This behavior is also present in the main branch and is not related to the optimization work.

<img width="1319" alt="Screenshot 2025-01-22 at 10 02 53 AM" src="https://github.com/user-attachments/assets/9eb32f5b-b0e8-415f-8701-5ee13a896820" />

andreitr commented:
Derive Testnet QA 

[Deployment](https://agora-next-derive-git-andrei-eng-508-delegate-fb7801-voteagora.vercel.app) 

The “Edit Delegate Statement” button remains disabled even after updating the statement. This behavior is also present in the main branch and is unrelated to the optimization work.
<img width="1334" alt="Screenshot 2025-01-22 at 10 44 08 AM" src="https://github.com/user-attachments/assets/618ea595-8b2a-4095-a20a-fd8dbd44fc52" />

andreitr commented:
Derive Testnet QA 

[Deployment](https://agora-next-derive-f64ya4wlb-voteagora.vercel.app/info) 

- create / update delegate [statement](https://agora-next-derive-f64ya4wlb-voteagora.vercel.app/delegates/0x1d671d1B191323A38490972D58354971E5c1cd2A)
- delegated voting power [txn](https://explorer-prod-testnet-0eakp60405.t.conduit.xyz/tx/0x9c3070a65df9cfef9bafd8cf1b97217529fa9935fffb34ee90e08f38c373257c)
- verified correct `delegated from` and `delegated to` table data

🚨 Observed a bug where the voting power is not displayed correctly on the delegate profile. It appears the cache is not being invalidated.

andreitr commented:
**ENS Testnet QA** 
[https://agora-next-9tr8tl60n-voteagora.vercel.app/](https://agora-next-9tr8tl60n-voteagora.vercel.app)
- updated delegate statement 
- delegated [txn](https://sepolia.etherscan.io/tx/0xbec854e8f29f35d8009a7052a56330260d812f6673b19e59f36d89527986881e)
- undelegated [txn](https://sepolia.etherscan.io/tx/0x644e639ff5027d2bb6af2bf5e88d386f35869f14a82008b9872ad29c1fb2ba52)
- created [proposal](https://agora-next-9tr8tl60n-voteagora.vercel.app/proposals/20291692341718005236250634318072028516645622919576176028162785918972025064407) 
- voted on proposal [txn](https://sepolia.etherscan.io/tx/0xcfc67b07b47d4bc5b2666eed24f40f489cd0cc5e06438977cb905f1bcabff4df)


**Uniswap Testnet QA**
[https://agora-next-uniswap-32kgq0eww-voteagora.vercel.app/](https://agora-next-uniswap-32kgq0eww-voteagora.vercel.app/)
- updated delegate statement 
- delegated [txn](https://sepolia.etherscan.io/tx/0x79d7fef18f456bc8010596ca95b80b6aa367b8476bc8004ce0dc17a76886e3a5)
- undelegated [txn](https://sepolia.etherscan.io/tx/0x473059e4df3bd394c1f8829ec84357d5adb5a3f6ccbdc587aa3579181682a694) 
- created a [proposal](https://agora-next-uniswap-32kgq0eww-voteagora.vercel.app/proposals/23) 
- voted [txn](https://sepolia.etherscan.io/tx/0x061f6b6e964f2beadbab3c9de0f49ebe024f837e8f3b780b5556d71c0a9951ef)


**Scroll Testnet QA**
[https://agora-next-scroll-b08axwmqg-voteagora.vercel.app/](https://agora-next-scroll-b08axwmqg-voteagora.vercel.app/)
- updated delegate statement 
- delegated [txn](https://scrollscan.com/tx/0xe648820145919c7e3403970faaca86ea548e60137e1d368cf163a998d84ea82d)
- created a proposal (verified locally against prod db)
- voted on a proposal (verified locally against prod db) 

**OP Testnet QA**
[https://agora-next-optimism-npqzjz265-voteagora.vercel.app/](https://agora-next-optimism-npqzjz265-voteagora.vercel.app/)
- updated delegate statement 
- verified loading of [large profiles](https://agora-next-optimism-npqzjz265-voteagora.vercel.app/delegates/0x2b888954421b424c5d3d9ce9bb67c9bd47537d12)
- basic delegation [txn](https://optimistic.etherscan.io/tx/0xd28502e0928021cf0da43ec8b3590e835de6017146eb6342f14e21f0f6e1d393)
- undelegate [txn](https://optimistic.etherscan.io/tx/0xa697323632531d5971455f30a57009e3f26f1bd3aa179ff87537478285c38149#eventlog)
